### DISCUSSION TITLE
Auto-generate all Async and Sync extensions of IManagementClient.

### DISCUSSION THREAD

**Discussion by inikulshin** [EasyNetQ/EasyNetQ.Management.Client#301] [Ideas]
Async - all those one-liners that accept `Vhost`, `Queue`, etc. parameters and call Async methods/extensions with `string` parameters.
Sync - all one-liners that call Async methods/extensions with exactly the same parameters continued with `.GetAwaiter().GetResult()`.

I already have a code that generates both (**83** Async and **163** Sync extensions), but I'm not sure what to do:
1. Push autogenerated extension files only and continue to maintain them manually.
2. Push autogenerated extension files and generator code in a separate solution to help to maintain extension files semi-automatically in rare(?) case of API changes.
3. Push generator code only as a part of main solution - however it'll require additional project and some ugly postbuild tasks...

**Comment by zidad** [comment]
Interesting! Are you not using a Roslyn generator with attributes, because then it should be possible to generate build time without needing any files at all? 

Alternatively, I was thinking this library could benefit from https://github.com/reactiveui/refit attributes on the interfaces, so we don't need any concrete implementation at all in our code base.

**Reply by inikulshin** [reply]
I used `Microsoft.CodeAnalysis.CSharp` for syntax tree generation, but I didn't know how to create generator right.
Now I'm sure it should be option (3) with source generator https://learn.microsoft.com/en-us/dotnet/csharp/roslyn-sdk/source-generators-overview.

**Reply by inikulshin** [reply]
@zidad I won't create any additional PR before my other PRs are closed.

**Reply by zidad** [reply]
That's fair enough! Sorry I'm a bit out of the loop in de EasyNetQ projects but I'll have a look now

**Comment by inikulshin** [comment]
Implemented in https://github.com/EasyNetQ/EasyNetQ.Management.Client/pull/304
